contracts/DXswapFeeReceiver.sol

```diff
@@ -6,122 +6,226 @@ import './interfaces/IWETH.sol';
+        (uint256 reserveIn, uint256 reserveOut) = token0 < token1 ? (reserve0, reserve1) : (reserve1, reserve0);
+        if (reserveIn == 0 || reserveOut == 0) return false;
+
+        uint256 priceImpact = amount.mul(10000) / reserveIn; // simplified formula
+        if (priceImpact > maxSwapPriceImpact) return false;
+
+        return true;
+    }
+
+    // Checks if LP has an extra external address which participates in the distrubution of protocol fee
+    // External Receiver address has to be defined and fee % > 0 to transfer tokens
+    function _splitAndTransferFee(
+        address pair,
+        address token,
+        uint256 amount
+    ) internal {
+        address _externalFeeReceiver = externalFeeReceivers[pair].externalReceiver;
+        uint32 _percentFeeToExternalReceiver = externalFeeReceivers[pair].feePercentage;
+
+        if (_percentFeeToExternalReceiver > 0 && _externalFeeReceiver != address(0)) {
+            uint256 feeToExternalReceiver = amount.mul(_percentFeeToExternalReceiver) / 10000;
+            uint256 feeToAvatarDAO = amount.sub(feeToExternalReceiver);
+            if (token == WETH) {
+                IWETH(WETH).withdraw(amount);
+                TransferHelper.safeTransferETH(_externalFeeReceiver, feeToExternalReceiver);
+                TransferHelper.safeTransferETH(ethReceiver, feeToAvatarDAO);
+            } else {
+                TransferHelper.safeTransfer(token, _externalFeeReceiver, feeToExternalReceiver);
+                TransferHelper.safeTransfer(token, fallbackReceiver, feeToAvatarDAO);
+            }
+        } else {
+            if (token == WETH) {
+                IWETH(WETH).withdraw(amount);
+                TransferHelper.safeTransferETH(ethReceiver, amount);
+            } else {
+                TransferHelper.safeTransfer(token, fallbackReceiver, amount);
+            }
+        }
+    }
+
+    // Convert tokens into ETH if possible, if not just transfer the token
+    function _takeTokenOrETH(
+        address pair,
+        address token,
+        uint256 amount
+    ) internal {
+        if (token != WETH && _isSwapPossible(token, WETH, amount)) {
+            // If it is not WETH and there is a direct path to WETH, swap tokens
+            uint256 amountOut = _swapTokensForETH(amount, token);
+            _splitAndTransferFee(pair, WETH, amountOut);
+        } else {
+            // If it is WETH or there is not a direct path from token to WETH, transfer tokens
+            _splitAndTransferFee(pair, token, amount);
+        }
     }
-    
+
     // Take what was charged as protocol fee from the DXswap pair liquidity
     function takeProtocolFee(IDXswapPair[] calldata pairs) external {
-        for (uint i = 0; i < pairs.length; i++) {
+        for (uint256 i = 0; i < pairs.length; i++) {
             address token0 = pairs[i].token0();
             address token1 = pairs[i].token1();
             pairs[i].transfer(address(pairs[i]), pairs[i].balanceOf(address(this)));
```

Comment:
IMHO we _need_ to perform some validation on the `pairs` elements here. Having external calls on unverified addresses makes me feel quite uneasy.